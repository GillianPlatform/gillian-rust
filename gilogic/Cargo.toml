[package]
name = "gilogic"
version = "0.1.0"
edition = "2021"

# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html

[dependencies]
gilogic_proc = { path = "../gilogic_proc" }
gilogic_proc_dummy = { path = "../gilogic_proc_dummy" }