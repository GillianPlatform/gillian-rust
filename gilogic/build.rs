fn main() {
    println!("cargo::rustc-check-cfg=cfg(gillian)");
}
