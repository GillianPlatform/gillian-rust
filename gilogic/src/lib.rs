#![allow(internal_features)]
#![feature(rustc_attrs)]
#![feature(never_type)]
#![feature(register_tool)]
#![feature(ptr_internals)]
#![feature(allocator_api)]
#![feature(tuple_trait, unboxed_closures)]
#![feature(stmt_expr_attributes, proc_macro_hygiene)]
#![register_tool(gillian)]

mod tys;

pub use tys::RustAssertion;

pub mod macros {
    pub use gilogic_proc::{
        assertion, assertion_test, borrow, close_borrow, extract_lemma, lemma, open_borrow,
        predicate, show_safety, specification,
    };
    pub mod prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_pcy as with_freeze_lemma_for_mutref;
    }
    pub mod no_prophecies {
        pub use gilogic_proc::with_freeze_lemma_for_mutref_no_pcy as with_freeze_lemma_for_mutref;
    }
}

mod seq;
pub use seq::Seq;

mod ownable;
pub use ownable::Ownable;
pub mod alloc;
pub mod iterated;
pub mod prophecies;

pub mod symex_ctrl {

    #[gillian::builtin]
    #[gillian::no_translate]
    #[rustc_diagnostic_item = "gillian::symex_ctrl::symex_next"]
    #[inline(always)]
    pub fn symex_next() {}
}

#[path = "stubs.rs"]
pub mod __stubs;
