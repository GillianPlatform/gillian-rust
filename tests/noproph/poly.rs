//@check-pass
extern crate gilogic;

use gilogic::{macros::show_safety, Ownable};

#[show_safety]
fn test<T: Ownable>(x: Box<T>) -> T {
    return *x;
}

#[show_safety]
fn test2() -> i32 {
    test(Box::new(12))
}
