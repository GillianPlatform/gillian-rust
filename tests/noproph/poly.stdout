#begin_init_data
{"Box":["Struct","ReprRust",[["0",["Adt",["Unique",[["Param",0]]]]],["1",["Param",1]]]],"Global":["Struct","ReprRust",[]],"NonNull":["Struct","ReprRust",[["pointer",["Ptr",{"mut":false,"ty":["Param",0]}]]]],"PhantomData":["Struct","ReprRust",[]],"Unique":["Struct","ReprRust",[["pointer",["Adt",["NonNull",[["Param",0]]]]],["_marker",["Adt",["PhantomData",[["Param",0]]]]]]]}
#end_init_data

import "i__binop.gil", "i__lang.gil", "i__std_shims.gil", "i__prophecies.gil", "i__std_shims_no_pcy.gil", "i__lang_no_pcy.gil";

axiomatic spec "test::<i32>"(x)
[[ "<std::boxed::Box<i32> as gilogic::Ownable>::own"(#x) * (x == #x) ]]
[["<i32 as gilogic::Ownable>::own"(ret)]]
normal

pred "<std::boxed::Box<T> as gilogic::Ownable>::own"(+pty_T0, +self):
  types(#lvar_0: Obj, #lvar_1: List)
  * ({{ {{ {{ {{ #lvar_0, #lvar_1 }} }}, {{  }} }}, {{  }} }} == #self)
  * <value>(#lvar_0, #lvar_1, pty_T0; #v)
  * "<T as gilogic::Ownable>::own"(pty_T0, #v)
  * (self == #self);


abstract pred "<T as gilogic::Ownable>::own"(+pty_T0, +self);


pred "<i32 as gilogic::Ownable>::own"(+self):
  (-2147483648i i<=# #self) * (#self i<=# 2147483647i) * types(#self: Int)
  * (self == #self);


pred "<std::boxed::Box<i32> as gilogic::Ownable>::own"(+self):
  types(#lvar_0: Obj, #lvar_1: List)
  * ({{ {{ {{ {{ #lvar_0, #lvar_1 }} }}, {{  }} }}, {{  }} }} == #self)
  * <value>(#lvar_0, #lvar_1, "i32"; #v)
  * "<i32 as gilogic::Ownable>::own"(#v)
  * (self == #self);


spec test(pty_T0, x)
[[ "<std::boxed::Box<T> as gilogic::Ownable>::own"(#pty_T0, #x) * (x == #x) * (pty_T0 == #pty_T0) ]]
[["<T as gilogic::Ownable>::own"(#pty_T0, ret)]]
normal

proc test(pty_T0, x) {
        m_1x := x;
        g_0 := m_1x;
        m_1x := [alloc]({{ "adt", "Box", {{ pty_T0, {{ "adt", "Global", {{  }} }} }} }});
        u := [store_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "adt", "Box", {{ pty_T0, {{ "adt", "Global", {{  }} }} }} }}, g_0);
  bb0:  g_2 := [load_value](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "adt", "Box", {{ pty_T0, {{ "adt", "Global", {{  }} }} }} }}, true);
        g_1 := l-nth(g_2, 0i);
        g_4 := [load_value](l-nth(l-nth(l-nth(l-nth(g_1, 0i), 0i), 0i), 0i), l-nth(l-nth(l-nth(l-nth(g_1, 0i), 0i), 0i), 1i), pty_T0, false);
        g_3 := l-nth(g_4, 0i);
        m_0 := g_3;
        goto bb1;
  bb1:  g_5 := m_0;
        ret := g_5;
        goto rlab;
  rlab: u := [free](l-nth(m_1x, 0i), l-nth(m_1x, 1i), {{ "adt", "Box", {{ pty_T0, {{ "adt", "Global", {{  }} }} }} }});
        return
};

spec test2()
[[ emp ]]
[["<i32 as gilogic::Ownable>::own"(ret)]]
normal

proc test2() {
  bb0:  g_0 := "std::boxed::Box::<T>::new"("i32", 12i);
        m_1 := g_0;
        goto bb1;
  bb1:  g_1 := m_1;
        g_2 := "test::<i32>"(g_1);
        m_0 := g_2;
        goto bb2;
  bb2:  g_3 := m_0;
        ret := g_3;
        goto rlab;
  rlab: return
};

